src/plugins/react.ts

```diff
@@ -0,0 +1,83 @@
+import { Plugin, PluginKey } from "prosemirror-state";
+
+/**
+ * This is a stand-in for the doc node itself, which doesn't have a
+ * unique position to map to.
+ */
+export const ROOT_NODE_KEY = Symbol("portal registry root key");
+
+export type NodeKey = string | typeof ROOT_NODE_KEY;
+
+/**
+ * Identifies a node view constructor as having been created
+ * by @nytimes/react-prosemirror
+ */
+export const REACT_NODE_VIEW = Symbol("react node view");
+
+export function createNodeKey() {
+  return Math.floor(Math.random() * 0xffffff).toString(16);
+}
+
+export const reactPluginKey = new PluginKey("@nytimes/react-prosemirror/react");
+
+/**
+ * Tracks a unique key for each (non-text) node in the
+ * document, identified by its current position. Keys are
+ * (mostly) stable across transaction applications. The
+ * key for a given node can be accessed by that node's
+ * current position in the document, and vice versa.
+ */
+export function react() {
+  return new Plugin({
+    key: reactPluginKey,
+    state: {
+      init(_, state) {
+        const next = {
+          posToKey: new Map<number, string>(),
```

Comment:
Is there a reason that we are using a `string` on line 36 and `NodeKey` on the following line?

Reply:
Oh interesting, yeah this is worth thinking about a bit more, maybe. `NodeKey` is just "`string` + `Symbol('portal rogistry root key')`", and technically we never actually put that root symbol in the plugin's bimap. The reason that `keyToPos` is typed as having `NodeKey` is just so that consumers don't have to do a special-case check to see whether they have the root symbol or a string before trying to get the position from the `keyToPos` map (that is, it's type-safe to do `pluginState.keyToPos.get(nodeKey)`, where `nodeKey` is of type `NodeKey`).

On the other hand, since we never actually put that root symbol in the bimap, technically it's correct that `pluginState.posToKey(pos)` always returns `string`, and you can use a `string` with any interface that accepts a `NodeKey` (since `NodeKey` is a subtype of `string`). So I think from a purely type-checking standpoint, this is "correct", but I also think it's somewhat unintuitive and might seem wrong to someone who's just using the API (which isn't _actually_ exposed to users, so this would just be "us in the future", but we should avoid confusing "us in the future", too!).

The only downside I can think of to typing `posToKey` as `Map<number, NodeKey>` is that you can technically take any result of `posToKey` and pass it to any function that expects a `string`, but if it was typed as `NodeKey`, Typescript wouldn't let you do that! I dunno what functions you'd be passing a node key to that weren't typed as NodeKey, so this seems pretty minor, as downsides go (and it's an internal API so it's not a big deal if we have to change it later).

Any thoughts after all that? Haha

Reply:
Also we should probably change the Symbol name! That was from a previous version of this plugin, and it doesn't really make sense here